src/v/pandaproxy/schema_registry/service.cc

```diff
@@ -84,10 +99,81 @@ class wrap {
         }
     }
 
+private:
+    inline net::unresolved_address
+    from_ss_sa(const ss::socket_address& sa) const {
+        return {fmt::format("{}", sa.addr()), sa.port(), sa.addr().in_family()};
+    }
+
+    security::audit::authentication::used_cleartext
+    is_cleartext(const ss::sstring& protocol) const {
+        return boost::iequals(protocol, "https")
+                 ? security::audit::authentication::used_cleartext::no
+                 : security::audit::authentication::used_cleartext::yes;
+    }
+    security::audit::authentication_event_options
+    make_authn_event_options(const server::request_t& rq) const {
+        return {
+          .auth_protocol = rq.user.sasl_mechanism,
+          .server_addr = from_ss_sa(rq.req->get_server_address()),
+          .svc_name = audit_svc_name,
+          .client_addr = from_ss_sa(rq.req->get_client_address()),
+          .is_cleartext = is_cleartext(rq.req->get_protocol_name()),
+          .user = {
+            .name = rq.user.name.empty() ? "{{anonymous}}" : rq.user.name,
+            .type_id = rq.user.name.empty()
+                         ? security::audit::user::type::unknown
+                         : security::audit::user::type::user}};
+    }
+    security::audit::authentication_event_options make_authn_event_error(
+      const server::request_t& rq,
+      const ss::sstring& username,
+      ss::sstring reason) const {
+        return {
+          .server_addr = from_ss_sa(rq.req->get_server_address()),
+          .svc_name = audit_svc_name,
+          .client_addr = from_ss_sa(rq.req->get_client_address()),
+          .is_cleartext = is_cleartext(rq.req->get_protocol_name()),
+          .user
+          = {.name = username, .type_id = security::audit::user::type::unknown},
+          .error_reason = reason};
+    }
+
+    void audit_authn_failure(
+      const server::request_t& rq,
+      const ss::sstring& username,
+      ss::sstring reason) const {
+        do_audit_authn(
+          rq, make_authn_event_error(rq, username, std::move(reason)));
+    }
+
+    void audit_authn(const server::request_t& rq) const {
+        do_audit_authn(rq, make_authn_event_options(rq));
+    }
+
+    void do_audit_authn(
+      const server::request_t& rq,
+      security::audit::authentication_event_options options) const {
+        vlog(
+          plog.trace, "Attempting to audit authn for {}", rq.req->format_url());
+        auto success = _audit_mgr.local().enqueue_authn_event(
```

Comment:
Please expose auth_mgr on service so that: `rq.context().service.audit_mgr()` can be used and all the wrapping noise can be dropped.

Reply:
would `rq.service().audit_mgr()` be acceptable?

Reply:
Even better!

Reply:
I should change the signature so it's `r.s().a()`